File: systems/controllers/residual_learner.cc
#include "residual_learner.h"

#include <utility>
#include <chrono>
using std::vector;
using drake::MatrixX;
using drake::systems::Context;

using Eigen::MatrixXd;
using Eigen::VectorXd;

namespace dairlib {
namespace systems {
using solvers::LCS;

namespace controllers {
ResidualLearner::ResidualLearner(){
  num_state = 19;
  num_velocity = 9;
  num_control = 3;
  num_lambda = 8;

  state_input_port_ =
      this->DeclareVectorInputPort(
              "x, u, t",
              OutputVector<double>(14, 13, 7))
          .get_index();

  lcs_output_port_ =
      this->DeclareAbstractOutputPort(
             "Residual LCS",
             &ResidualLearner::CalcResidual)
      .get_index();
}

void ResidualLearner::CalcResidual(const Context<double>& context,
                                      LCS* residual_lcs) const {

  // get lcm_robot_output values
  auto robot_output = (OutputVector<double>*)this->EvalVectorInput(context, state_input_port_);
  double timestamp = robot_output->get_timestamp();

  int N = 1;
  MatrixXd A = 0 * MatrixXd::Identity(num_velocity, num_state);
  MatrixXd B = 0 * MatrixXd::Identity(num_velocity, num_control);
  MatrixXd D = 0 * MatrixXd::Identity(num_velocity, num_lambda);
  VectorXd d = 0 * VectorXd::Ones(num_velocity);

  MatrixXd E = 0 * MatrixXd::Identity(num_lambda, num_state);
  MatrixXd F = 0 * MatrixXd::Identity(num_lambda, num_lambda);
  MatrixXd H = 0 * MatrixXd::Identity(num_lambda, num_control);
  VectorXd c = 0 * VectorXd::Ones(num_lambda);

  LCS res_lcs(A, B, D, d, E, F, H, c, N);
  *residual_lcs = res_lcs;
}

}
}
}

File: examples/franka_trajectory_following/legacy/residual_learner.py
# pydrake
from pydrake.systems.framework import Context, LeafSystem
from pydrake.common.value import AbstractValue

# pydairlib
from pydairlib.solvers.lcs import LCS
from pydairlib.systems.framework import OutputVector

# others
import numpy as np

class ResidualLearner(LeafSystem):
    def __init__(self):
        """
            Constructor for the ResidualLearner, currently fix the dimension of the problem,
            can be modified in the future to make it linked with external problem settings
        """
        LeafSystem.__init__(self)

        # problem dimensions
        self.num_state = 19
        self.num_velocity = 9
        self.num_control = 3
        self.num_lambda =12

        # port number definition
        # input port
        self.state_input_port = self.DeclareVectorInputPort("x,u,t",OutputVector(14,13,7)).get_index()
        # output port
        self.lcs_output_port = self.DeclareAbstractOutputPort("Residual LCS", lambda: AbstractValue.Make(LCS()),
                                                              self.CalcResidual).get_index()
        OutputPort = self.get_output_port()
        print(OutputPort.get_data_type())


    def CalcResidual(self, context:Context, residual_lcs:LCS):
        # get data from input port
        robot_output = self.EvalVectorInput(context, self.state_input_port)
        timestamp = robot_output.get_timestamp()

        N = 1
        A = np.zeros((self.num_velocity, self.num_state))
        B = np.zeros((self.num_velocity, self.num_control))
        D = np.zeros((self.num_velocity, self.num_lambda))
        d = np.zeros((self.num_velocity,1))

        E = np.zeros((self.num_lambda, self.num_state))
        F = np.zeros((self.num_lambda, self.num_lambda))
        H = np.zeros((self.num_lambda, self.num_control))
        c = np.zeros((self.num_lambda, 1))

        residual_lcs = LCS(A, B, D, d, E, F, H, c, N)





